src/logics/Theory.cc

```diff
@@ -39,37 +39,35 @@ Theory::SubstitutionResult Theory::computeSubstitutions(const PTRef fla)
     PTRef root = fla;
     // l_True : exists and is valid
     // l_False : exists but has been disabled to break symmetries
-    Map<PTRef,PtAsgn,PTRefHash> allsubsts;
+    MapWithKeys<PTRef,PtAsgn,PTRefHash> allsubsts;
     // This computes the new unit clauses to curr_frame.units until closure
     while (true) {
         // update the current simplification formula
         PTRef simp_formula = root;
-        Map<PTRef,lbool,PTRefHash> new_units;
-        vec<Map<PTRef,lbool,PTRefHash>::Pair> new_units_vec;
+        MapWithKeys<PTRef,lbool,PTRefHash> new_units;
         vec<PtAsgn> current_units_vec;
         // Get U_i
         getLogic().getNewFacts(simp_formula, new_units);
         // Add the newly obtained units to the list of all substitutions
         // Clear the previous units
-        new_units.getKeysAndVals(new_units_vec);
-        for (int i = 0; i < new_units_vec.size(); i++) {
-            Map<PTRef,lbool,PTRefHash>::Pair unit = new_units_vec[i];
-            current_units_vec.push(PtAsgn(unit.key, unit.data));
+        const auto & new_units_vec = new_units.getKeys();
+        for (PTRef key : new_units_vec) {
+            current_units_vec.push(PtAsgn{key, new_units[key]});
         }
-        Map<PTRef,PtAsgn,PTRefHash> newsubsts;
+
+        MapWithKeys<PTRef,PtAsgn,PTRefHash> newsubsts;
         lbool res = getLogic().retrieveSubstitutions(current_units_vec, newsubsts);
         getLogic().substitutionsTransitiveClosure(newsubsts);
         if (res != l_Undef)
             root = (res == l_True ? getLogic().getTerm_true() : getLogic().getTerm_false());
         PTRef new_root = Substitutor(getLogic(), newsubsts).rewrite(root);
         bool cont = new_root != root;
         // remember the substitutions
-        vec<Map<PTRef,PtAsgn,PTRefHash>::Pair> newsubsts_vec;
-        newsubsts.getKeysAndVals(newsubsts_vec);
-        for (int i = 0; i < newsubsts_vec.size(); ++i) {
-            PTRef key = newsubsts_vec[i].key;
-            if (!allsubsts.has(key) && newsubsts_vec[i].data.sgn == l_True) {
-                allsubsts.insert(key, newsubsts_vec[i].data);
+        auto & newsubsts_vec(newsubsts.getKeys());
+        for (PTRef key : newsubsts_vec) {
+            const auto & el(newsubsts[key]);
```

Comment:
```suggestion
            const auto & el = newsubsts[key];
```
Since it is `PtAsgn`, which is a small object I would maybe even prefer not use a reference here.

Reply:
We can do that.